inspire_utils/date.py

```diff
@@ -0,0 +1,233 @@
+        # XXX: this only handles dates after 1000, which should be sufficient
+        formatted = ('{:02d}'.format(part) for part in non_empty)
+        date = '-'.join(formatted)
+
+        return date
+
+    @classmethod
+    def parse(cls, date, **kwargs):
+        """Parse a date given in arbitrary format.
+
+        This attempts to parse the input date, given in an arbitrary format
+
+        Args:
+            date(str): date to normalize
+            **kwargs: these are passed to the `dateutil.parser.parse` function
+                that is used internally to parse the date. Most notably,
+                `yearfirst` and `datefirst` flags can be used if the ordering of
+                the date parts is know.
+
+        Returns:
+            PartialDate: an object holding the parsed date.
+
+        Raises:
+            ValueError: when the date cannot be parsed or no year is present in
+                it.
+
+        Examples:
+            >>> PartialDate.parse('30 Jun 1686')
+            'PartialDate(year=1686, day=6, month=30)'
+        """
+        # In order to detect partial dates, parse twice with different defaults
+        # and compare the results.
+        default_date1 = datetime.datetime(1, 1, 1)
+        default_date2 = datetime.datetime(2, 2, 2)
+
+        parsed_date1 = parse_date(date, default=default_date1, **kwargs)
+        parsed_date2 = parse_date(date, default=default_date2, **kwargs)
+
+        has_year = parsed_date1.year == parsed_date2.year
+        has_month = parsed_date1.month == parsed_date2.month
+        has_day = parsed_date1.day == parsed_date2.day
+
+        if has_year:
+            year = parsed_date1.year
+        else:
+            raise ValueError('date does not contain a year')
+        month = parsed_date1.month if has_month else None
+        day = parsed_date1.day if has_day else None
+
+        return cls(year, month, day)
+
+    @classmethod
+    def from_parts(cls, year, month=None, day=None):
+        """Build a PartialDate from its parts.
+
+        Unlike the standard constructor, the parts don't have to be `int` s but
+        can be strings containing textual month information.
+
+        Examples:
+            >>> PartialDate.from_parts('1686', 'June', '30')
+            PartialDate(year=1686, month=6, day=30)
+        """
+        # XXX: 0 is not a valid year/month/day
+        non_empty = itertools.takewhile(
+            bool, (str(part) if part else None for part in (year, month, day))
+        )
+        return cls.parse(u'-'.join(non_empty), yearfirst=True)
+
+    def pprint(self):
+        """Pretty print the date.
+
+        Examples:
+            >>> PartialDate(1686, 6, 30).pprint()
+            u'Jun 30, 1686'
+        """
+        if not self.month:
+            return dates.format_date(datetime.date(self.year, 1, 1), 'YYYY', locale='en')
+        if not self.day:
+            return dates.format_date(datetime.date(self.year, self.month, 1), 'MMM, YYYY', locale='en')
```

Comment:
the locale is probably not needed in Inspire, but it was needed on my laptop that is set-up in Italian.